src/shady-css/tokenizer.js

```diff
@@ -0,0 +1,229 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+import { matcher } from './common';
+import { Token, boundaryTokenTypes } from './token';
+
+const currentToken = Symbol('currentToken');
+const nextToken = Symbol('nextToken');
+
+/**
+ * Class that implements tokenization of significant lexical features of the
+ * CSS syntax.
+ */
+class Tokenizer {
+  /**
+   * Create a Tokenizer instance.
+   * @param {string} cssText The raw CSS string to be tokenized.
+   */
+  constructor(cssText) {
+    this.cssText = cssText;
+    this.offset = 0;
+    this[currentToken] = null;
+  }
+
+  /**
+   * The current token that will be returned by a call to `advance`. This
+   * reference is useful for "peeking" at the next token ahead in the sequence.
+   * If the entire CSS text has been tokenized, the `currentToken` will be null.
+   * @type {Token}
+   */
+  get currentToken() {
+    if (this[currentToken] == null) {
+      this[currentToken] = this[nextToken]();
+    }
+
+    return this[currentToken];
+  }
+
+  /**
+   * Advance the Tokenizer to the next token in the sequence.
+   * @return {Token} The current token prior to the call to `advance`, or null
+   * if the entire CSS text has been tokenized.
+   */
+  advance() {
+    let token;
+    if (this[currentToken] != null) {
+      token = this[currentToken];
+      this[currentToken] = null;
+    } else {
+      token = this[nextToken]();
+    }
+    return token;
+  }
+
+  /**
+   * Extract a slice from the CSS text, using two tokens to represent the range
+   * of text to be extracted. The extracted text will include all text between
+   * the start index of the first token and the offset index of the second token
+   * (or the offset index of the first token if the second is not provided).
+   * @param {Token} startToken The token that represents the beginning of the
+   * text range to be extracted.
+   * @param {Token} endToken The token that represents the end of the text range
+   * to be extracted. Defaults to the startToken if no endToken is provided.
+   * @return {string} The substring of the CSS text corresponding to the
+   * startToken and endToken.
+   */
+  slice(startToken, endToken) {
```

Comment:
That might be useful, but the tokens are not linked so the start and end tokens won't give you sufficient information without changes to their structure. Lazy text production sounds like a good optimization, but maybe we should do it in a future iteration.